### PR TITLE
Create bring-your-ideas-to-world-with-kubectl-plugins.md

### PR DESCRIPTION
**Pull request by corneliusweig**
Hey, I'm co-maintainer of the SIG-CLI subproject [krew](krew.dev). We want to publish a small series of blog posts to promote the development of kubectl plugins. This is the first part :)

fyi @ahmetb

### CODE REVIEW

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+---
+title: Bring your ideas to world with Kubectl Plugins
```
Comment (sftim) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]: I'd write both “kubectl” and “plugins” lowercase:
```suggestion
title: Bring your ideas to world with kubectl plugins
```

**Comment by corneliusweig** [comment]
Hey @sftim, thanks for reviewing! It's a great idea to also link to the docs on kubernetes.io. At your convenience, have another look.

**Comment by sftim** [comment]
:ok_hand: 

I'll leave a formal review to the blog team.

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+---
+title: Bring your ideas to world with kubectl plugins
```
Comment (sftim) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]: Oh, one thing, sorry.
```suggestion
title: Bring your ideas to the world with kubectl plugins
```

**Comment by ahmetb** [comment]
Hello, blog team. 👋 Do you have any estimated timelines for this?
This is the part 1 of a 3-post series we're starting.
It would be good to know how long does it take to publish a post so we can pace them.

**Comment by corneliusweig** [comment]
Hey @parispittman, do you know how we can get this blog post reviewed? It's been ready for review for quite a while now :blush:

**Comment by mrbobbytables** [comment]
Sorry, this fell through the cracks -- I'm poking the other blog team members to get folks on board.

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+title: Bring your ideas to the world with kubectl plugins
+date: tbd
+---
+
+**Author:** Cornelius Weig (TNG Technology Consulting GmbH)
+
+`kubectl` is the single most important tool to interact with Kubernetes. As
+such, it has to fulfill the needs of a plethora of user personas, each with
+their own needs and opinions. One way to make `kubectl` do what you need is to
+build the new functionality into `kubectl`.
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
`kubectl` is the most critical tool to interact with Kubernetes and has to address multiple user personas, each with their own needs and opinions. 
One way to make `kubectl` do what you need is to build new functionality into `kubectl`.
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+## Challenges with building commands into `kubectl`
+
+However, that's easier said than done. Being such an important cornerstone of
+Kubernetes, any meaningful change to `kubectl` needs to undergo a Kubernetes
+Enhancement Proposal (KEP) where the intended change is discussed beforehand.
+
+When it comes to implementation, you'll find that `kubectl` is an ingenious and
+complex piece of engineering. It might take a long time to get used to
+processes and style of the codebase to get done what you want to achieve. Next
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
the processes and style of the codebase to get done what you want to achieve. Next
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+the requirements of the Kubernetes maintainers -- after all, they need to take
+over ownership of this feature and maintain it from the day it's merged.
+
+When everything goes well, you can finally rejoice. Your code will be shipped
+with the next Kubernetes release. Well, that could mean you need to wait
+another 3 months to ship your idea in `kubectl` if you are unlucky.
+
+So this was the happy path where everything goes well. But there are good
+reasons why your new functionality may never make it into `kubectl`. For one,
+`kubectl` has a particular look and feel and violating that style will not be
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
 `kubectl` has a particular look and feel, and violating that style will not be 
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+of users, the maintainers may simply reject your proposal as `kubectl` needs to
+address common needs.
+
+But this doesn’t mean you can’t ship your ideas to `kubectl` users.
+
+## What if you didn’t have to change `kubectl` to add functionality?
+
+This is where `kubectl` [plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) shine.
+Since `kubectl` v1.12, you can simply
+drop executables into your `PATH` which follow the naming pattern
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
drop executables into your `PATH`, which follows the naming pattern 
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+`kubectl-myplugin`. Then you can execute this plugin as `kubectl myplugin` and
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
`kubectl-myplugin`. Then you can execute this plugin as `kubectl myplugin`, and
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+it will just feel like a normal sub-command of `kubectl`.
+
+Plugins give you the opportunity to try out new experiences like terminal UIs,
+colorful output, specialized functionality, or other innovative ideas. You can
+go creative, as you’re the owner of your own plugin.
+
+Further, plugins offer safe experimentation space for commands you’d like to
+propose to `kubectl`. By pre-releasing as a plugin, you can push your
+functionality faster to the end users and quickly gather feedback. For example,
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
functionality faster to the end-users and quickly gather feedback. For example, 
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+## How to get started with developing plugins
+
+If you already have an idea for a plugin, how do you best make it happen?
+First you have to ask yourself if you can implement it as a wrapper around
+existing `kubectl` functionality. If so, writing the plugin as a shell script
+is often the best way forward, because the resulting plugin will be small,
+works cross-platform, and has a high level of trust because it is not
+compiled.
+
+On the other hand, if the plugin logic is complex, a general purpose language
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
On the other hand, if the plugin logic is complex, a general-purpose language 
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+is usually better. The canonical choice here is Go, because you can use the
+excellent `client-go` library to interact with the Kubernetes API. The Kubernetes
+maintained [sample-cli-plugin](https://github.com/kubernetes/sample-cli-plugin)
+demonstrates some best practices and can be used as a template for new plugin
+projects.
+
+When the development is done, you just need to ship your plugin to the
+Kubernetes users. For best plugin installation experience and discoverability,
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
Kubernetes users. For the best plugin installation experience and discoverability, 
```

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+you should consider doing so via the
+[krew](https://github.com/kubernetes-sigs/krew) plugin manager. For an in-depth
+discussion about the technical details around `kubectl` plugins refer to the
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]:
```suggestion
discussion about the technical details around `kubectl` plugins, refer to the 
```

**Comment by corneliusweig** [comment]
Oh wow, thanks for taking a look @onlydole!
All your suggestions should be included now. :confetti_ball:

**Comment by mrbobbytables** [comment]
From a technical standpoint, everything looks good 👍

**Comment by corneliusweig** [comment]
Thanks @mrbobbytables!
Do you have a rough estimate how long such PRs usually take? Just asking because we want to follow up with at least two more.

**Comment by mrbobbytables** [comment]
probably 48hrs on the long side, just need more sign off and update the dates.

File: content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
+---
+title: Bring your ideas to the world with kubectl plugins
+date: tbd
```
Comment (onlydole) [inline content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]: Howdy @corneliusweig - this is the only hold out piece for the post and then we can LGTM it 😄

Comment (corneliusweig) [reply content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]: Oh sorry, now this notification slipped through at my end :confused:

If I understand the process correctly, the release date will be determined by the blog team, right?

Comment (mrbobbytables) [reply content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]: Usually it is, we'll edit it as the last step before going live.

Comment (mrbobbytables) [reply content/en/blog/_posts/bring-your-ideas-to-world-with-kubectl-plugins.md]: I've added a date for tomorrow.

**Pull request merged**